include/rapidcheck/state/Commands.hpp

```diff
@@ -49,5 +196,81 @@ void showValue(const std::vector<
   }
 }
 
+template <typename CommandType>
+struct ParallelCommands {
+  using Cmd = CommandType;
+  using Cmds = Commands<CommandType>;
+
+  ParallelCommands() {}
+
+  ParallelCommands(const Cmds &prefix, const Cmds &left, const Cmds &right)
+      : prefix(prefix)
+      , left(left)
+      , right(right) {}
+
+  Cmds prefix;
+  Cmds left;
+  Cmds right;
+};
+
+template <typename Cmd, typename Model, typename Sut>
+void runAllParallel(const ParallelCommands<Cmd> &commands, const Model &state, Sut &sut) {
+
+  // Verify pre-conditions for all possible interleavings
+  detail::isValidSequence(commands, state);
```

Comment:
Also, is it up to `runAllParallel` to verify the fact that it is a valid sequence or is it expected? Let's say that we want the generator to only generate valid parallel sequences, that means that it has to verify this internally and then we would verify this here once again. Something to think about since verification is not free.